### DISCUSSION TITLE
CSF referencing raises error: index exceeds number of array elements

### DISCUSSION THREAD

**Discussion by fprinse** [kschan0214/sepia#90] [Q&A]
Hi,

I'm running the SEPIA pipeline but in the dipole inversion step I get an error regarding index exceeding the number of array elements with using CSF as a reference tissue:

> Index exceeds the number of array elements. Index must not exceed 1.
> 
> Error in extract_CSF (line 43)
>     numPixels_sorted(i);
> 
> Error in QSMMacro (line 92)
>                 mask_ref    = extract_CSF(r2s,mask,voxelSize)>0;
> 
> Error in SepiaIOWrapper (line 322)
> [chi,mask_ref] = QSMMacro(localField,mask_QSM,matrixSize,voxelSize,algorParam,headerAndExtraData);
> 
> Error in sepiaIO (line 79)
>     SepiaIOWrapper(input,output,maskFullName,algorParam);
> 
> Error in sepia_config (line 38)
> sepiaIO(input,output_basename,mask_filename,algorParam);
> 
> Error in run (line 91)
> evalin('caller', strcat(script, ';'));
> 
> Error in sepia>PushbuttonStart_Callback (line 350)
>     run(configFilename);
>  
> Error while evaluating UIControl Callback.

I provided 4D phase and magnitude images acquired on a Siemens 7T. I combined the echoes using MATLAB. I haven't had this error previously using the same configuration
![image](https://github.com/user-attachments/assets/fe80c5da-46c4-4318-8e8f-4689c23f38a1)

Thanks!

**Comment by kschan0214** [comment]
Hi, sorry for the late follow up! Have you been able to find a solution yet? It seems to me that something went wrong with the brain mask but I cannot be sure without checking the input data.

**Reply by tsalo** [reply]
I have stumbled across the same error with 5-echo 3T data. I've overlaid my brain mask on the magnitude images and it looks fine to me. I can share the data if necessary.

**Reply by kschan0214** [reply]
Hi Taylor, it would be great if you could share the data and the script (sepia_config.m) file with me. I can then try to reproduce the error and get back to you.

**Reply by tsalo** [reply]
Thanks. I've uploaded the working directory, which includes the input NIfTIs, the header file, and the script, to https://upenn.box.com/s/5sq8anm0kmercf7x7ojyrx7iufwba90i. I can't use the MATLAB GUI on my university's server, so I've had to set variables based on the descriptions in the SEPIA documentation.

EDIT: The inputs are called `python_mag.nii`, `python_phase.nii`, and `python_mask.nii`.

**Reply by kschan0214** [reply]
Thanks for providing the data! 

I managed to reproduce the error. It was caused by the incorrect unit of the 'TE' variable in the header file (millisecond was used instead of the default unit of second), making the threshold values used in extract_CSF.m inappropriate and failed to segment the CSF. This can be solved by converting the TE values in the header to ms (i.e. ```TE = TE*1e-3;```) and re-save the header file. Another unrelated issue is the 'B0' variable is in 'uint8' type which will casue error in the processing. You can simply convert the 'B0' datatype to 'double' (i.e. ```B0 = double(B0);```) to solve the issue. You should be able to run the pipeline once these two issues are resolved. I will also update SEPIA to handle this bug in the furture release. 

Let me know if you still have any troubles!

**Reply by tsalo** [reply]
Thanks so much! This resolved my issue.
